File: assistant/assistant.py
import json
from openai import OpenAI
import os
assistant = [{'id': 'asst_GW0wpM1ZAeeUhOdpBFxBIC5r', 'created_at': 1716200659, 'description': None, 'instructions': 'You are a personal math tutor. Answer questions briefly, in a sentence or less.', 'metadata': {}, 'model': 'gpt-4-1106-preview', 'name': 'Math Tutor', 'object': 'assistant', 'tools': [], 'response_format': 'auto', 'temperature': 1.0, 'tool_resources': {'code_interpreter': None, 'file_search': None}, 'top_p': 1.0}]
def show_json(obj):
    print(json.loads(obj.model_dump_json()))

client = OpenAI(api_key=os.environ.get("OPENAI_API_KEY", "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"))

def createAssistant():
    assistant = client.beta.assistants.create(
        name="Math Tutor",
        instructions="You are a personal math tutor. Answer questions briefly, in a sentence or less.",
        model="gpt-4-1106-preview",
    )
    show_json(assistant)

def further():
    thread = client.beta.threads.create()
    show_json(thread)

    # Add messages
    message = client.beta.threads.messages.create(
    thread_id=thread.id,
    role="user",
    content="I need to solve the equation `3x + 11 = 14`. Can you help me?",
    )
    show_json(message)

    run = client.beta.threads.runs.create(
    thread_id=thread.id,
    assistant_id=assistant.id,
    )
    show_json(run)

further()

File: tts/assistant.js
// tts/assistant.js

import OpenAI from "openai";
import dotenv from "dotenv";
import { readFileAsync } from "./async_read.js";

dotenv.config({ path: "../.env" });

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

export async function createAssistant(role) {
  let additionalData = "";

  if (role === "teacher") {
    try {
      additionalData = await readFileAsync("./tts/Mr_Ranedeer.txt"); // Adjust the path as necessary
    } catch (err) {
      console.error("Failed to read teacher data:", err);
    }
  }

  const instructions =
    role === "teacher"
      ? `You are a teacher. Provide educational content and answer should be in a format "[teacher]: your speech.Use the following to enhance learning style ${additionalData}"`
      : `You are a student. repeat the last message just said and tell a joke about it by laughing and answer should be of format "[student] : your answer"`;

  const assistant = await openai.beta.assistants.create({
    model: "gpt-3.5-turbo",
    name: `${role.charAt(0).toUpperCase() + role.slice(1)} Assistant`,
    instructions: instructions,
  });

  return assistant;
}

export async function generateResponse(assistant, conversationHistory, prompt) {
  const thread = await openai.beta.threads.create({
    messages: [...conversationHistory, { role: "user", content: prompt }],
  });

  const stream = await openai.beta.threads.runs.create(thread.id, {
    assistant_id: assistant.id,
    stream: true,
  });

  let accumulatedText = "";
  for await (const event of stream) {
    if (event.event === "thread.message.delta") {
      const chunk = event.data.delta.content?.[0];
      if (chunk && "text" in chunk && chunk.text.value) {
        accumulatedText += chunk.text.value;
      }
    }
  }

  conversationHistory.push({ role: "assistant", content: accumulatedText });
  return { accumulatedText, conversationHistory };
}
